### ISSUE TITLE
rename 'master' branch to 'main'

### ISSUE THREAD

**Issue by aiuto** [bazelbuild/bazel#12200]
Because we should

**Comment by philwo** [comment]
Do you have experience with and guidance on how to do this without breaking all the tools and scripts that currently have the old branch name hardcoded? I acknowledge the importance of this, but I'm a bit at a loss on how to do this technically without breaking everything and everyone.

As a rough estimate of the work involved:

grep over our CI repo:

<img width="1179" alt="Screen Shot 2020-09-30 at 23 18 16" src="https://user-images.githubusercontent.com/504652/94740597-3b8ab300-0373-11eb-8dfd-7917cb037120.png">

Plus the Google-internal stuff, the config in the Buildkite pipeline settings, the gitsync container, the Copybara setup, the Gerrit configuration / ACLs, pending pull requests, ...

This has to be very, very well coordinated.

**Comment by meisterT** [comment]
See https://github.com/github/renaming for some guidance, specifically the mention that this will be more seamless later this year.

**Comment by aiuto** [comment]
I had seen that page before, but not the "seamless" renaming coming later.  I did this with rules_pkg a month ago and learned some pain points
- old PRs are against master.  If you don't remember to rebase them off main, they go to the wrong branch.
- copybara pulls from 'master' by default, not from the default branch. Easy to fix, but the copy.bara.sky files all need update
- CI seems to work without change.

**Comment by UebelAndre** [comment]
Hey, would someone be able to elaborate on:

> Because we should

?

**Comment by meisterT** [comment]
> Hey, would someone be able to elaborate on:
> 
> > Because we should

The term `master` is offensive to some people. There's nothing special about the name in how git treats branches it was just historically the name for the first branch.

**Comment by philwo** [comment]
Hey @UebelAndre :)

Terms like "master / slave" or "whitelist / blacklist" are considered as offensive by many people (e.g. reminding them of slavery or racism). Obviously the *intent* of using these terms was not to provoke these negative feelings - they just somehow happened to become standard terms in the IT, but if there's a good neutral term available to replace them, I think it's just kind to consider doing so. Thus, there has been a recent movement to find better words for what we try to say! Here's some background: https://www.zdnet.com/article/github-to-replace-master-with-alternative-term-to-avoid-slavery-references/.

Often the recommended replacements actually convey the intended meaning better than the historically used words - some examples:
- It's the `main` branch, because it's our... wait for it... main branch, where the development happens and release branches are cut from. :)
- A "CI worker machine" is called like that, because it works on CI jobs.
- "allowlist" and "denylist" directly say what they're about (allowing or denying access) rather than relying on cultural background to understand "white" as "good" and "black" as "bad". Even if one doesn't see racism in the original words, the replacements are easier to understand for non-native speakers and people from cultures where this whole "white vs. black" idea doesn't exist.

**Comment by UebelAndre** [comment]
@meisterT @philwo Thanks guys! This ticket makes me very happy 😄

**Comment by cherryland** [comment]
All these "inclusive language" requests show, how the toxic sewage of the American culture is spilling over.

Wherever I am, the stinking Americans are trying to push their poisonous obsession of black slaves to the rest of the world, trying to enslave the world with their guilt.

**The word master is akin to maestro. In sports, the German word Meister is used for the current national, European or world champion (e.g. Deutscher Meister, Europameister, Weltmeister).**

I hope all Americans kill themselves in their passion for hatred. Go out Americans, slit open the throats of your neighbours, do it in the name of social justice, but let the rest of the world in peace. We are fed up with your bullshit.

**Comment by cherryland** [comment]
Today it is the word "master", tomorrow it will be numerical digit "1", because "1" it is less inclusive than the numerical digit "2" and so on.

I hope China takes over soon so we can get rid of all the Americans, and their toxic poison. Let them kill themselves.

**Comment by cherryland** [comment]
Today it is the word "master", tomorrow it will be numerical digit "1", because "1" it is less inclusive than the numerical digit "2" and so on.

I hope China takes over soon, so we can get rid of all the Americans and their toxic poison. Let them kill themselves in the American streets, but let the rest of the world in peace. Get your national problems out of the source code. We have had it, enough is enough.

**Comment by sventiffe** [comment]
@cherryland your offensive behavior and your wording is not acceptable. We're open to discussion in a civilised manner but you've clearly crossed the line by wishing others death and by insulting people. 

We are dedicated to inclusiveness and will continue our effort on this issue. Your disagreement is noted, but your messages disqualify you from any further involvement.

**Comment by cherryland** [comment]
@sventiffe your offensive behavior and your wording is not acceptable. We're open to discussion in a civilised manner but you've clearly crossed the line by wishing others death and by insulting people.

We are dedicated to inclusiveness and will continue our effort on this issue. Your disagreement is noted, but your messages disqualify you from any further involvement.

**Comment by jmillikin** [comment]
(This is mostly a bikeshed comment, feel free to ignore if it would complicate the migration)

Would it be possible for the new name to be `trunk` rather than `main`? That's what the primary development branch was usually called before Git was released, and it's easier to grep for than `main` because it doesn't conflict with the C/C++ default entry point.

**Comment by aiuto** [comment]
I'll see the bikeshed, and raise you.  trunk always disappointed me as
a name.
If we want to be precise about it, there are only branches. We simply are
designating one as more central/important than the others.
We reflect that in the way we do development. Forking a branch gives us
something that exists independently.
The trunk does not feed the branches, as in a real tree. We have to do
extraordinary things to pull code from the primary
development branch to any other branch.  I would rather see 'dev" or 'core'
as the name over trunk. That captures the fact
that we are doing the ongoing development there.

**Comment by UebelAndre** [comment]
Related to this, I think https://docs.bazel.build/versions/3.2.0/skylark/config.html#user-defined-transitions and all the relevant components should have counterparts made with updated names.

**Comment by UebelAndre** [comment]
I've noticed that quite a few repositories in the `bazelbuild` org that haven't had any updates relevant to this GH issue. Instead of having them pop up in each repo. I think it'd mean a lot to have someone comb through them and make related updates. It would also be nice to see this ticket get actioned.

**Comment by aiuto** [comment]
@philwo   Many of the last few repositories (website, blog, and bazel itself) are going to require coordinated changes to production pipelines.  Can you free up someone who knows the push pipelines for a few hours sometime in December to help out on this? I suspect that for every repository except bazel we can just do the rename at any time and not worry about people's pending PRs.  For bazel, we should have a more coordinated rollover period so that everyone does not have to scramble to rebased their existing work.  

There are notes about the steps needed in the fixit doc from earlier this year. I think they cover the process - except for some of the production things I might not know about.

**Comment by brentleyjones** [comment]
Not stale.

**Comment by aiuto** [comment]
@meteorcloudy  We *could* do this, but it's just going to take the willpower to declare a day when people can't commit changes for an hour or two while we update the copybara scripts.  How about July 3 or 4? No one in the US will be working, so that will reduce volume.

**Comment by meteorcloudy** [comment]
Good point, but I guess first we should have a rough estimation on who and what needs to be updated after the renaming and do a clear communication in advance with all impacted parties.

**Comment by aiuto** [comment]
IIRC, from the last time I did this:

   - Rename on github is now easy.
   - The copybara push to GOB needs to change from master to main
   - The GOB step needs to change.  I have no idea what happens here.
   - I think all users must do something with PRs in flight.  That should
   come with a cookbook example
   - People will find it easier if they rename the branches in their
   personal clones. that needs an example
